src/sbuf/SBuf.h

```diff
@@ -580,28 +451,50 @@ class SBuf
     /// std::string export function
     std::string toStdString() const { return std::string(buf(),length()); }
 
-    const_iterator begin() const {
-        return const_iterator(*this, 0);
-    }
-
-    const_iterator end() const {
-        return const_iterator(*this, length());
-    }
-
-    const_reverse_iterator rbegin() const {
-        return const_reverse_iterator(*this, length());
-    }
+    /// SBuf object identifier meant for test cases and debugging.
+    /// Does not change when object does, including during assignment.
+    const InstanceId<SBuf> id;
 
-    const_reverse_iterator rend() const {
-        return const_reverse_iterator(*this, 0);
+    /* std::string API */
```

Comment:
1) That `/* std::string API */` line refers to the entire block of methods following it. The placement is correct.

2) No these two operators do not CoW.

3) The two assignment methods moved to here perform data-replacement on `*this` directly equivalent to how `std::string` could be implemented. Providing the same set of guarantees and non-guarantees.

Reply:
> 1. That `/* std::string API */` line refers to the entire block of methods following it. The placement is correct.

A single violation of the label inside the block invalidates the idea of labeling the entire block. One cannot correctly label a bag of apples "green apples" if one of the apples in the bag is red. The placement is incorrect.

> 2. No these two operators do not CoW.

```C++
SBuf& operator =(const SBuf &S) { return assign(S); } ///< \sa SBuf::assign
```

The SBuf assignment operator (the subject of this change request) does CoW -- assign(SBuf) does not create a copy of the assigned SBuf S, as a "normal" assignment operator would do. Instead, it reuses S's storage for its own needs, incrementing the reference counter. Avoiding copies on such S "reads" is what copy-on-write is about!

> 3. The two assignment methods moved to here perform data-replacement on `*this` directly equivalent to how `std::string` could be implemented. Providing the same set of guarantees and non-guarantees.

Again, std::string assignment cannot do CoW since C++11. The change was driven by [N2534](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2008/n2534.html) which says "This change disallows copy-on-write implementations". I have already linked to a higher-level [discussion](https://stackoverflow.com/q/12199710) confirming the same C++11 outcome.